### PR TITLE
fix(PerformanceManager): Fix regression in hard coding card0 for gpu settings.

### PR DESCRIPTION
**Pull request by pastaq**
- Update powertools script to use cardX in /sys/class/drm passed from PerformanceManager.
- Update Plaform to detect all cards in the system. We default to the first entry for simplicity in this hotfix but we need to find a way to link the pci data to the devices in vulkaninfo.
- Update Platform to use native methods of determining the current display device vendor, model, and driver.
- Fix bug in PowerTools menu that would cause GPU clocks to be set to 200Mhz on startup and override that value in the profile.
- Fix Intel TDP and GPUCLK settings.

### CODE REVIEW

File: core/global/platform.gd [hunk 1]
```diff
@@ -388,80 +392,179 @@ func _get_lscpu_info() -> Array:
...
+	# to look into vulkaninfo as this can only gather the data
+	# for the currently active GPU. Vulkaninfo can provide data
+	# on all detected GPU devices.
+	match RenderingServer.get_video_adapter_vendor():
+		"AMD", "AuthenticAMD", 'AuthenticAMD Advanced Micro Devices, Inc.', "Advanced Micro Devices, Inc. [AMD/ATI]":
+			gpu_info.vendor = "AMD"
+		"Intel", "GenuineIntel", "Intel Corporation":
+			gpu_info.vendor = "Intel"
+		"Nvidia":
+			gpu_info.vendor = "Trash" # :D
```
Comment (ShadowApex) [inline core/global/platform.gd:411]: lolol

File: core/global/platform.gd [hunk 2]
```diff
@@ -388,80 +392,179 @@ func _get_lscpu_info() -> Array:
...
+	# TODO: This is bad and lazy. We need to identify a correct way to link the
+	# PCI Address to the rendering info
+	# Assume the first card is the one we are using.
+	cards = _get_gpu_cards()
+	if cards.size() <= 0:
+		logger.warn("GPU Data could not be derived.")
+		return null
+
+	logger.debug("Assuming " + cards[0].name + " is the correct card!")
+	gpu_info.card = cards[0]
```
Comment (ShadowApex) [inline core/global/platform.gd:430]: It looks like we can parse `glxinfo` to get the device ID of the currently used device:

image

`0x1636` is the device ID of my integrated graphics.

If we can pull this device ID out, we can match on `card.device` for the correct `Card` instance.

**Review by ShadowApex — CHANGES_REQUESTED** [review]

**Comment by pastaq** [comment]
> why not use /dev/dri/by-id?

/dev/dri only has character files that represent the device, but don't have the subdirectories that contain vendor and device information.

File: core/global/platform.gd (hunk 2 above)
Comment (pastaq) [reply core/global/platform.gd:430]: Bet

Comment (ShadowApex) [reply core/global/platform.gd:430]: A function like this should work to parse that device id:

```gdscript

func get_current_card_device() -> String:
	var cmd := Command.new("glxinfo")
	if await cmd.execute() != OK:
		return ""
	
	for line in cmd.stdout.split("\n"):
		if not "Device: " in line:
			continue
		
		# Match on the last part of the string to get the device ID
		# E.g. Device: AMD Radeon Graphics (renoir, LLVM 16.0.6, DRM 3.54, 6.5.5-arch1-1) (0x1636)
		var regex := RegEx.new()
		regex.compile("0x(\\d+)\\)$")
		var result := regex.search(line)
		if not result:
			continue
		var match_str := result.get_string()

		return match_str.replace("0x", "").replace(")", "")

	return ""
```

Comment (ShadowApex) [reply core/global/platform.gd:430]: We should probably grab the vendor too to cleanly match on vendor + device:

```
    Vendor: AMD (0x1002)
    Device: AMD Radeon Graphics (renoir, LLVM 16.0.6, DRM 3.54, 6.5.5-arch1-1) (0x1636)
```

File: core/global/platform.gd [hunk 3]
```diff
@@ -66,12 +71,13 @@ func _init() -> void:
...
 
 	var flags := get_platform_flags()
-	
+
 	# Set hardware platform provider
 	if PLATFORM.ABERNIC_GEN1 in flags:
 		platform = load("res://core/platform/handheld/abernic/abernic_gen1.tres")
 	if PLATFORM.ALLY_GEN1 in flags:
 		platform = load("res://core/platform/handheld/asus/rog_ally_gen1.tres")
+		@warning_ignore("unsafe_property_access")
```
Comment (ShadowApex) [inline core/global/platform.gd:80]: Instead of ignoring this unsafe property access, we should do a type check/cast so this is a safe call.

File: core/global/platform.gd [hunk 4]
```diff
@@ -147,7 +154,7 @@ func load(root: Window) -> void:
 func get_handheld_gamepad() -> HandheldGamepad:
 	if not platform:
 		return null
-	return await platform.get_handheld_gamepad()
+	return platform.get_handheld_gamepad()
```
Comment (ShadowApex) [inline core/global/platform.gd:157]: Why was this `await` removed? Seems unrelated to all the other changes and maybe the source of the handheld issues you saw?

File: core/global/platform.gd [hunk 5]
```diff
@@ -374,94 +385,303 @@ func _read_cpu_info() -> CPUInfo:
...
+
+	return [card_vendor_name, card_device_name, card_subdevice_name]
+
+
+func _get_cards_from_vulkan() ->Array[PackedStringArray]:
+	var vulkan_cards: Array[PackedStringArray] = []
+	var args = ["-c", "vulkaninfo", "--summary"]
+	var output: Array = _do_exec("bash", args)
+	var exit_code = output[1]
+	if exit_code:
```
Comment (ShadowApex) [inline core/global/platform.gd:609]:
```suggestion
	if exit_code != OK:
```

File: core/global/platform.gd [hunk 6]
```diff
@@ -374,94 +385,303 @@ func _read_cpu_info() -> CPUInfo:
...
+
+## Returns the string of the currently active GPU
+func get_active_gpu_device() -> PackedStringArray:
+	var vendor: String
+	var device: String
 	var args = ["-c", "glxinfo", "-B"]
 	var output: Array = _do_exec("bash", args)
 	var exit_code = output[1]
+
 	if exit_code:
```
Comment (ShadowApex) [inline core/global/platform.gd:642]:
```suggestion
	if exit_code != OK:
```

File: core/global/platform.gd [hunk 7]
```diff
@@ -489,10 +710,21 @@ class CPUInfo extends Resource:
...
+	func _to_string() -> String:
+		return "<CPUInfo:" \
+			+ " Vendor: (" + str(vendor) \
+			+ ") Model: (" + str(model) \
+			+ ") Core count: (" + str(core_count) \
+			+ ") Cores Enabled: (" + str(cores_available) \
+			+ ") Boost Capable: (" + str(boost_capable) \
+			+ ") SMT Ccapable: (" + str(smt_capable) \
+			+ ")>"
+
```
Comment (ShadowApex) [inline core/global/platform.gd:722]: Nice!!

File: core/systems/performance/performance_manager.gd [hunk 8]
```diff
@@ -562,7 +565,7 @@ func _set_sane_defaults() -> void:
 ## Reads the pp_od_clk_voltage from sysfs and returns the OD_RANGE values. This file will 
 ## be empty if not in "manual" for pp_od_performance_level.
 func _read_amd_gpu_clock_limits() -> void:
-	var args := ["/sys/class/drm/card0/device/pp_od_clk_voltage"]
+	var args := ["/sys/class/drm/" + gpu.card.name + "/device/pp_od_clk_voltage"]
```
Comment (ShadowApex) [inline core/systems/performance/performance_manager.gd:568]: Is it possible for this to ever be null? Do we have a fallback path?

File: rootfs/usr/share/opengamepadui/scripts/powertools [hunk 9]
```diff
@@ -8,65 +8,79 @@ fi
 
 ## CPU Generics
 cpuBoost(){
-  /usr/bin/echo $1 > /sys/devices/system/cpu/cpufreq/boost
-}
-
-cpuPower(){
-  /usr/bin/cpupower $1 $2 $3
+ state=$1
+  /usr/bin/echo $state > /sys/devices/system/cpu/cpufreq/boost
```
Comment (ShadowApex) [inline rootfs/usr/share/opengamepadui/scripts/powertools:12]: :tada:

File: rootfs/usr/share/opengamepadui/scripts/powertools [hunk 10]
```diff
@@ -8,65 +8,79 @@ fi
...
 amdGpuClock(){
-  /usr/bin/echo "s 0 ${1}" > /sys/class/drm/card0/device/pp_od_clk_voltage
-  /usr/bin/echo "s 1 ${2}" > /sys/class/drm/card0/device/pp_od_clk_voltage
-  /usr/bin/echo "c" > /sys/class/drm/card0/device/pp_od_clk_voltage
+  min_clk=$1
+  max_clk=$2
+  card=$3
+  /usr/bin/echo "s 0 ${min_clk}" > /sys/class/drm/${card}/device/pp_od_clk_voltage
+  /usr/bin/echo "s 1 ${max_clk}" > /sys/class/drm/${card}/device/pp_od_clk_voltage
+  /usr/bin/echo "c" > /sys/class/drm/${card}/device/pp_od_clk_voltage
```
Comment (ShadowApex) [inline rootfs/usr/share/opengamepadui/scripts/powertools:33]: This change would break any users that are using the update zip, but on an older AUR/systemd-ext version. Can we detect and set a default value for `card`?

E.g.
```bash
if [[ "$3" != "" ]]; then
    card=$3
else
    if [ -L /sys/class/drm/card0 ]; then
        card="card0"
    else
        card="card1"
    fi
fi
```

File: rootfs/usr/share/opengamepadui/scripts/powertools [hunk 11]
```diff
@@ -8,65 +8,79 @@ fi
...
-  /usr/bin/echo $1 > /sys/class/drm/card0/device/power_dpm_force_performance_level
+  mode=$1
+  card=$2
+  /usr/bin/echo $mode > /sys/class/drm/${card}/device/power_dpm_force_performance_level
 }
 
 setAmdPerfLevelWrite() { # set power_dpm_force_performace_level write
-  /usr/bin/chmod a+w /sys/class/drm/card0/device/power_dpm_force_performance_level
+  card=$1
+  /usr/bin/chmod a+w /sys/class/drm/${card}/device/power_dpm_force_performance_level
```
Comment (ShadowApex) [inline rootfs/usr/share/opengamepadui/scripts/powertools:55]: Same as above comment, we should set a default for `card` so we don't break update zip users

File: rootfs/usr/share/opengamepadui/scripts/powertools (hunk 10 above)
Comment (pastaq) [reply rootfs/usr/share/opengamepadui/scripts/powertools:33]: No it won't, the card argument will just get ignored in that case. Adding logic to a file to handle a case that only exists because you can't edit that file makes no sense.

File: core/global/platform.gd [hunk 12]
```diff
@@ -159,17 +166,20 @@ func get_platform_flags() -> Array[PLATFORM]:
...
-	var product_name := _read_sys("/sys/devices/virtual/dmi/id/product_name")
-	logger.debug("Product Name: " + product_name)
 	return product_name
 
 
 ## Returns the hardware vendor name
 func get_vendor_name() -> String:
-	var vendor_name := _read_sys("/sys/devices/virtual/dmi/id/sys_vendor")
-	logger.debug("Vendor Name: " + vendor_name)
 	return vendor_name
```
[the remark this answers is not in the document]
Comment (pastaq) [reply core/global/platform.gd:183]: What if we checked for an empty string? Third information is static so I'm trying to reduce the number of expensive blocking calls

File: rootfs/usr/share/opengamepadui/scripts/powertools (hunk 9 above)
Comment (pastaq) [reply rootfs/usr/share/opengamepadui/scripts/powertools:12]: ?

File: core/global/platform.gd (hunk 12 above)
[the remark this answers is not in the document]
Comment (ShadowApex) [reply core/global/platform.gd:183]: If you want to save on the number of calls, then you can just set the class variables directly at the top of the file with a call to the functions:

```gdscript
var product_name := get_product_name()
var vendor_name := get_vendor_name()
```

File: rootfs/usr/share/opengamepadui/scripts/powertools (hunk 9 above)
Comment (ShadowApex) [reply rootfs/usr/share/opengamepadui/scripts/powertools:12]: Just a fan of using variable names.

File: rootfs/usr/share/opengamepadui/scripts/powertools (hunk 10 above)
Comment (ShadowApex) [reply rootfs/usr/share/opengamepadui/scripts/powertools:33]: You're right. I had this backwards. I was thinking of a case where this new script would be in place but with an older version of OpenGamepadUI. We shouldn't ever run into this problem.

File: core/global/platform.gd (hunk 4 above)
Comment (pastaq) [reply core/global/platform.gd:157]: It's not, I removed it after the issues arose. I was cleaning up warnings from the Godot editor to tidy up the PR and it mentioned await was not necessary here as its not a co-routine..

File: core/global/platform.gd (hunk 12 above)
[the remark this answers is not in the document]
Comment (pastaq) [reply core/global/platform.gd:183]: get_product_name and get_vendor_name are used all over the place. This wont reduce any calls.

[the remark this answers is not in the document]
Comment (ShadowApex) [reply core/global/platform.gd:183]: I only saw those in a few places, but what if instead we do:

```gdscript
var _vendor_name: String

## Returns the hardware vendor name
func get_vendor_name() -> String:
    if _vendor_name == "":
        _vendor_name = _read_sys("/sys/devices/virtual/dmi/id/sys_vendor")
    return _vendor_name
```

File: core/global/platform.gd [hunk 13]
```diff
@@ -374,94 +385,303 @@ func _read_cpu_info() -> CPUInfo:
...
+			logger.debug("Found subdevice_name: " + card_subdevice_name)
+			break
+
+	return [card_vendor_name, card_device_name, card_subdevice_name]
+
+
+func _get_cards_from_vulkan() ->Array[PackedStringArray]:
+	var vulkan_cards: Array[PackedStringArray] = []
+	var args = ["-c", "vulkaninfo", "--summary"]
+	var output: Array = _do_exec("bash", args)
```
[the remark this answers is not in the document]
Comment (pastaq) [reply core/global/platform.gd:607]: I tried this initially. The output was getting truncated for some reason. Seems to be working now so I'll revert.

File: core/global/platform.gd [hunk 14]
```diff
@@ -374,94 +385,303 @@ func _read_cpu_info() -> CPUInfo:
...
+		i += 1
+	return vulkan_cards
+
+
+## Returns the string of the currently active GPU
+func get_active_gpu_device() -> PackedStringArray:
+	var vendor: String
+	var device: String
 	var args = ["-c", "glxinfo", "-B"]
 	var output: Array = _do_exec("bash", args)
```
[the remark this answers is not in the document]
Comment (pastaq) [reply core/global/platform.gd:639]: SAA

File: core/global/platform.gd [hunk 15]
```diff
@@ -374,94 +385,303 @@ func _read_cpu_info() -> CPUInfo:
...
+
 	if exit_code:
+		logger.error("Failed to run glxinfo")
 		return []
-	return  output[0][0].split("\n") as Array
-	
+
+	var stdout := (output[0][0] as String).split("\n") as Array
+	for line in stdout:
+		if not "Device: " in line and not "Vendor: " in line:
```
[the remark this answers is not in the document]
Comment (pastaq) [reply core/global/platform.gd:648]: That won't work. If either case is false this will trigger. We want it to only trigger if both cases are false.

File: core/systems/performance/performance_manager.gd [hunk 16]
```diff
@@ -437,7 +437,7 @@ func _change_cpu_cores():
 func _enable_performance_write() -> void:
 	match gpu.vendor:
 		"AMD":
-			await _async_do_exec(POWERTOOLS_PATH, ["pdfpl", "write"])
+			await _async_do_exec(POWERTOOLS_PATH, ["pdfpl", "write", gpu.card.name])
```
[the remark this answers is not in the document]
Comment (pastaq) [reply core/systems/performance/performance_manager.gd:440]: No, but the call will fail either way.

File: core/systems/performance/performance_manager.gd (hunk 8 above)
Comment (pastaq) [reply core/systems/performance/performance_manager.gd:568]: We could set it to card1 by default. If that is incorrect though it won't work anyway and we have a bigger issue that detection didn't work.

File: rootfs/usr/share/opengamepadui/scripts/powertools (hunk 9 above)
Comment (pastaq) [reply rootfs/usr/share/opengamepadui/scripts/powertools:12]: Yeah, I conceded after I had to remember what each of these did specifically.

File: core/global/platform.gd (hunk 3 above)
Comment (pastaq) [reply core/global/platform.gd:80]: I did, it ignores me and still has the warning. I think there is a bug though because even this shows a warning in the editor.

File: core/global/platform.gd (hunk 13 above)
[the remark this answers is not in the document]
Comment (ShadowApex) [reply core/global/platform.gd:607]: If you do ever need to use bash, it should be in the form of: 

`_do_exec("bash", ["-c", "vulkaninfo --summary"])`

instead of

`_do_exec("bash", ["-c", "vulkaninfo", "--summary"])`

File: core/global/platform.gd (hunk 15 above)
[the remark this answers is not in the document]
Comment (ShadowApex) [reply core/global/platform.gd:648]: You're right, my bad!

File: core/global/platform.gd (hunk 12 above)
[the remark this answers is not in the document]
Comment (pastaq) [reply core/global/platform.gd:182]: Yeah that's what I had in mind

**Review by ShadowApex — APPROVED** [review]
